✨Implemented Placement-Decision Logic

vMaroon opened the pull request:
## Summary
Implemented PlacementDecision calculation and handling logic, alongside a few fixes
 
The implementation:
- Fixes a race-condition that was present in the use of gvkMapper, while also introducing gvr-mapping
- Adds a supporting struct: PlacementDecisionResolver that maintains the resolutions of all placement-decisions
- Adds handling for placement-decision objects, which includes handling out-of-sync objects and creating/updating objects when a decision is changed
- Marks code-pieces to be deleted when Nir's transport code is ready

The flow is as follows:
An object changes -> placement-decision resolutions are updated -> modified placement-decisions are enqueued to be synced by workers -> new spec is calculated and patched for each updated resolution.

### Important Notes
~~- This PR SHOULD NOT BE MERGED~~
~~- Singleton status is not yet addressed #1680 ~~
~~- The pluggable transport #1588 needs to be enabled #1692~~
~~- The code DOES NOT disable @pdettori's mixed implementation of placement-transport~~
~~- Only after changes are reviewed and the above points are resolved, I will delete marked code~~
Update after review:
- We will merge in 2 steps: this PR inserts code that maintains PlacementDecisions, another PR will enable the pluggable transport.

- **This code inherits the flow and styling existing in the current code**
  - **Please refrain from reviewing existing code through this PR. Comments on non-changed code should be addressed in a separate issue**
     - Related code refactoring issues:
       - #1690
       - #1683 
       - #1682 
       - #1679
       - #1677 

## Related issue(s)
#1520

MikeSpreitzer commented:
This PR should include the corresponding updates to the placement controller's description in the architecture document.

MikeSpreitzer commented:
Are there any important distinctions among the 40 commits here now?

MikeSpreitzer commented on pkg/crd/crds.go:
```diff
@@ -43,7 +43,10 @@ import (
 
 // CRDs to apply
-var crdNames = map[string]bool{"placements.edge.kubestellar.io": true}
+var crdNames = map[string]bool{
```
What is the `bool` for?

Why not use https://github.com/kubernetes/apimachinery/blob/v0.28.2/pkg/util/sets/set.go#L31 ?

MikeSpreitzer commented on pkg/util/gvk-gvr-mapper.go:
```diff
@@ -0,0 +1,104 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// GvkGvrMapper is a thread-safe mapping between GVKs and GVRs.
```
The comments here do not even imagine that there might be a question of whether the relation between GVK and GVR is 1:1. The comments here do not say anything --- exclusive or inclusive --- about subresources.

vMaroon replied:
I didn't refer to these cases in this version, but it is doable without affecting the rest of the code since it is interfaced. Do you think this should be updated as part of this PR? or commented as TODO?

MikeSpreitzer replied:
Not sure what you mean by "all cases". I think updating this one comment is sufficient. Just state that this mapper is only for top-level resources and there is an expectation that, within that restricted scope, the relation between resource and Kind is 1:1.

MikeSpreitzer commented on pkg/util/gvk-gvr-mapper.go:
```diff
@@ -0,0 +1,104 @@
+// GvkGvrMapper is a thread-safe mapping between GVKs and GVRs.
+type GvkGvrMapper interface {
+	// Add adds a mapping between the given GVK and GVR.
+	Add(gvk schema.GroupVersionKind, gvr schema.GroupVersionResource)
```
What if, just before this call, the given gvk was associated with a different gvr? And vice-versa?

MikeSpreitzer replied:
(A) The reader should be told what happens.

(B) Are you sure that you have made the right choice for what happens?

vMaroon replied:
(A) Added

(B) After adding the statement you suggested in the above comment, this scenario "is not supported", no?

MikeSpreitzer replied:
It is not clear what "This overwrites the GVK <-> GVR mapping if pre-existing" means. Since there is nothing to a mapping but a mapping, what would overwriting a mapping with itself even be? A no-op?

The question is, what happens if this method is invoked at a time when the given gvk is already associated with a _different_ gvr. Also, what if the given gvr is already associated with a _different_ gkv?

MikeSpreitzer replied:
I just added a comment to #1703, including my expectation that the planned revision will not completely remove the need for mapping between "kind" and "resource". I expect the mapper abstraction to remain. The mapper abstraction's definition should say something about the issue in the original comment of this conversation. It might say "the mapper panics if ..." or it might say that any previous association with either the gvk or gvr is removed or it might be redefined to return an error in the case of partial overlap.

MikeSpreitzer commented on pkg/util/gvk-gvr-mapper.go:
```diff
@@ -0,0 +1,104 @@
+	Add(gvk schema.GroupVersionKind, gvr schema.GroupVersionResource)
+	// DeleteByGvkKey deletes the mapping associated with the given GVK.
+	// The key format is that outputted by utils.KeyForGroupVersionKind.
+	DeleteByGvkKey(string)
```
Why not make the entire interface use the upstream structure (instead of the local string encoding)?

vMaroon replied:
Add/delete of mappings is done by the crd handling code. For convenience, deletion is done by the key as string since that is what's present when used.

I agree it would be better if the delete functions used the upstream structure, but that would require either:
1. Writing a reverse function that extracts GVK from the key, while handling empty group and assuming validity.
2. Refactoring the crd handling code, and the mappings in the controller, which need to be done regardless because they suffer from possible race conditions.

I prefer option 2, which is a part of refactoring the placement code as a whole that will follow later.

What do you think?

MikeSpreitzer replied:
Sure, that would be OK.

MikeSpreitzer commented:
This commits in this branch include many violations of https://docs.google.com/document/d/1iwkhpoV76-J-3G-Rif4B1cmL9SWBO_f_fE6WK8jvsDc

MikeSpreitzer commented on pkg/placement/controller.go:
```diff
@@ -390,12 +412,17 @@ func (c *Controller) reconcile(ctx context.Context, key util.Key) error {
+				obj = util.EmptyUnstructuredObjectFromKey(key)
+			} else if errors.IsNotFound(err) {
+				// The resource no longer exist, which means it has been deleted.
+				c.logger.Info("object in work queue no longer exists",
```
s/in/referenced from/

MikeSpreitzer commented on pkg/placement/controller.go:
```diff
@@ -413,13 +440,24 @@ func (c *Controller) reconcile(ctx context.Context, key util.Key) error {
 			return err
 		}
+	} else if util.IsPlacementDecision(obj) {
+		if err := c.handlePlacementDecision(obj); err != nil {
+			return err
+		}
+		return nil
```
This could be simpler.
```suggestion
		return c.handlePlacementDecision(obj)
```

MikeSpreitzer commented on pkg/placement/controller.go:
```diff
@@ -446,13 +484,13 @@ func (c *Controller) reconcile(ctx context.Context, key util.Key) error {
-	lister := c.listers[key.GvkKey]
-	if lister == nil {
-		utilruntime.HandleError(fmt.Errorf("could not get lister for key: %s", key.GvkKey))
+	pLister := c.listers[key.GvkKey()]
```
What does the "p" stand for here?

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,173 @@
+)
+
+// handlePlacementDecision syncs a placement-decision object with what is resolved by the placement-resolver.
+func (c *Controller) handlePlacementDecision(obj runtime.Object) error {
```
`Controller` also has a `handleObject` method. The reader wonders why PlacementDecision objects need a special handler.

The problem here is using the word "handle" to mean two things: handle an event notification from an informer, and do the sync AKA reconcile work on an object reference pulled from the work queue. I strongly recommend using different words for these two in the func names. For example, this method might be called `syncPlacementDecision`.

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,173 @@
+
+// handlePlacementDecision syncs a placement-decision object with what is resolved by the placement-resolver.
+func (c *Controller) handlePlacementDecision(obj runtime.Object) error {
+	placementDecision, err := runtimeObjectToPlacementDecision(obj.DeepCopyObject())
```
Why the extra deep copy here? `runtimeObjectToPlacementDecision` does a deep copy, right?

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,173 @@
+func (c *Controller) handlePlacementDecision(obj runtime.Object) error {
+	placementDecision, err := runtimeObjectToPlacementDecision(obj.DeepCopyObject())
+	if err != nil {
+		return fmt.Errorf("failed to handle placement-decision: %w", err)
```
s/handle placement-decision/convert from Unstructured to PlacementDecision/

MikeSpreitzer commented on pkg/placement/controller.go:
```diff
@@ -413,13 +440,24 @@ func (c *Controller) reconcile(ctx context.Context, key util.Key) error {
 		if err := c.handleCRD(obj); err != nil {
 			return err
 		}
+	} else if util.IsPlacementDecision(obj) {
```
Note that the if condition could be based on `key` rather than `obj`. If this `if` were moved to happen before the complex and buggy logic for workload objects then we would not the hack of an `isBeingCreated` field in the Key; syncing a reference to a PlacementDecision demonstrably needs no more than the Key. A work queue should only hold pure object references, not also instructions about an intended state change, because the relevant state change may be different by the time the object reference is dequeued for syncing.

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,173 @@
+	placementDecisionSpec *v1alpha1.PlacementDecisionSpec) error {
+	unstructuredPlacementDecision, err := placementDecisionSpecToUnstructuredObject(pd, placementDecisionSpec)
+	if err != nil {
+		return fmt.Errorf("failed to update or create placement decision: %w", err)
```
Think about what the user will see when this error is wrapped at line 55.

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,173 @@
+// If the object already exists, it is updated. Otherwise, it is created.
+func (c *Controller) updateOrCreatePlacementDecision(pd *v1alpha1.PlacementDecision,
+	placementDecisionSpec *v1alpha1.PlacementDecisionSpec) error {
+	unstructuredPlacementDecision, err := placementDecisionSpecToUnstructuredObject(pd, placementDecisionSpec)
```
The name `placementDecisionSpecToUnstructuredObject` tells the reader that it is just a conversion. But that is not true; this func also side-effects `pd`. Better to rename or, better yet, factor that part out to appear right here.

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,173 @@
+	}
+
+	// set owner reference
+	if err = setOwnerReference(c.dynamicClient, unstructuredPlacementDecision); err != nil {
```
Creating `placementDecisionSpec` must have required reading the `Placement` object. Why make `setOwnerReference` read it again?

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,173 @@
+				return fmt.Errorf("failed to create placement decision: %w", err)
+			}
+
+			c.logger.Info("created placement decision", "name", pd.GetName())
```
I do not think we want to log _both_ "created" and "updated".

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,173 @@
+	return list, nil
+}
+
+func runtimeObjectToPlacementDecision(obj runtime.Object) (*v1alpha1.PlacementDecision, error) {
```
Since this does not work for _any_ `runtime.Object` that represents a PlacementDecision, but only ones using `*unstructured.Unstructured`, I would name this func more specifically. Maybe `unstructuredRuntimeObjectToPlacementDecision`. But more likely I would do the cast from `runtime.Object` to `*unstructured.Unstructured` higher in the stack, leaving this func only to convert from `*unstructured.Unstructured` to `PlacementDecision`.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+const clusterScopedNamespace = ""
```
Why not use https://github.com/kubernetes/apimachinery/blob/v0.28.2/pkg/apis/meta/v1/types.go#L280 ?

vMaroon replied:
@MikeSpreitzer 
All comments until this point were implemented (unless I left a question in comment) in commit [2d3373040a3530d9e4e2f491cc807a68c34cf43f](https://github.com/kubestellar/kubestellar/pull/1681/commits/2d3373040a3530d9e4e2f491cc807a68c34cf43f).

All comments below this point (even if now marked outdated) - are not addressed yet.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+	sync.RWMutex
+
+	// map key is GVK/namespace/name as outputted by util.key.GvkNamespacedNameKey()
+	objectIdentifierToKeyMap map[string]*util.Key
```
(A) What goes wrong if you drop "Map" from this field's name?

(B) The point of this field is to hold a set of Key, right? Why not use `Set[Key]` (importing https://github.com/kubernetes/apimachinery/blob/v0.28.2/pkg/util/sets/set.go#L24).

vMaroon replied:
(A) is fine by me but I dont think this is critical.

(B) unless I explicitly implement Key comparison, the field DeletedObject would be problematic. Therefore I wonder if it's a necessary change.

MikeSpreitzer replied:
I see that you made change (A). Regarding (B), I do not follow your remark about implementing Key comparison. The go language does not allow user-defined comparison and hashing behavior for its built-in `map`s (I sure wish it did!).  As long as a a Key is not a pure object reference, the way to avoid string hackery would be to ensure that every domain value in the map has no extraneous content. One approach would be to define `func clean(Key) Key` that removes the DeletedObject if present, and apply `clean` before indexing the map. Another approach would be to factor Key into a pure (clean) part plus the Deleted object, and use only the pure part for indexing the map.  For example:

```go
type ObjectRef struct {
	GVK            schema.GroupVersionKind
	NamespacedName cache.ObjectName
}
type Key struct {
	ObjectRef
	DeletedObject  *runtime.Object
}
```

vMaroon replied:
If I understood correctly, in B you suggest I switch to Set[Key] instead of map[string]*Key, right? To use Key in Set, it would need to implement Comparable, which it does, since all its fields are Comparable. BUT, the problem is two Keys can refer to the same object, with DeletedObject changing, thus making them different keys.

The latter is not a problem in the current implementation because I use an object-unique string to map Keys.

I agree that Key must be reworked, and perhaps a lot of the "buggy" flow can be discarded when we delete the transport code from the placement controller. But I think it should be in the followup refactoring work I referred to in a previous comment. What do you think?

MikeSpreitzer replied:
I wrote two ways that you could have a map indexed by values where the `DeletedObject` field is not an issue, while still having Paolo's logic and transport code.

This is an FYI, does not have to be done in this PR. Can indeed be done after removing the transport code.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+
+// updateDestination updates the destinations list in the resolution.
+// This function is thread-safe.
+func (resolution *placementResolution) updateDestinations(destinations []string) {
```
FYI, I would call this `setDestinations`.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+	return exists
+}
+
+// updateDestination updates the destinations list in the resolution.
```
typo here: plural lost.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+
+	// the following optimize the building of the workload by maintaining pointers to the object-wrapper structs
+	// that are key'd by GVR
+	ClusterScopeDownsyncObjectsMap := map[schema.GroupVersionResource]*v1alpha1.ClusterScopeDownsyncObjects{}
```
Why the initial cap on this name?

MikeSpreitzer commented on pkg/util/gvk-gvr-mapper.go:
```diff
@@ -0,0 +1,104 @@
+	DeleteByGvrKey(string)
+	// GetGvr returns the GVR associated with the given GVK if it exists.
+	// The returned boolean indicates whether it exists or not.
+	GetGvr(gvk schema.GroupVersionKind) (*schema.GroupVersionResource, bool)
```
Why return a _pointer to_ the GVR instead of the GVR itself? It seems to me that returning the pointer just adds work.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+	// Since namespaceScopeDownsyncObjectsMap groups objects by NS, these maps are used to efficiently locate the
+	// * struct by GVR
+	// * objects slice by namespace (for GVR)
+	namespaceScopeDownsyncObjectsMap := map[schema.GroupVersionResource]*v1alpha1.NamespaceScopeDownsyncObjects{}
+	nsObjectsLocationInSlice := map[string]int{} // key is GVR/ns to avoid 2D maps
```
Why use a pointer (to a slice member) in one case and an index into the slice in the other case? Both work, I do not see a functional problem here; just wondering.

vMaroon replied:
Most likely a certain chain of thoughts at the time of writing this optimization led to this seeming more natural than using a pointer. I don't mind either.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+	matchesCounter := 0
+
+	// fill up matches
+	for _, destination := range placementDecisionDestinations {
+		matches[destination.ClusterId] = nil
+	}
```
Using https://github.com/kubernetes/apimachinery/blob/v0.28.2/pkg/util/sets/set.go#L31 :

```suggestion
	matches := New[v1alpha1.Destination](placementDecisionDestinations...)
```

Then do length check for early out.
If that passes then iterate over destinations and look for `v1alpha1.Destination{ClusterID: destination}` in `matches`.

vMaroon replied:
Length check is done first thing in this function, above the marked code.
The use of Set is fine

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+
+	// calculate match by counting matching objects then comparing to length.
+	// this works since entries are unique.
+	matches := map[string]interface{}{}
```
`PlacementDecision` objects will be easier for users to consume if the lists are kept in alphabetical order.

vMaroon replied:
I dont understand the correlation between this comment and marked code

MikeSpreitzer replied:
It is somewhat distant. It was triggered by thinking about more efficiently comparing slices. But that is moot, now the comparison is between a map and a slice.

Even so, the thought about consistently ordering the slice contents has merit on its own. When users go looking for changes, they will be easier to find if there is consistent ordering in the slices.

MikeSpreitzer replied:
I opened #1716 for the ordering.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+
+	// map key is GVK/namespace/name as outputted by util.key.GvkNamespacedNameKey()
+	objectIdentifierToKeyMap map[string]*util.Key
+	destinations             []string
```
Wouldn't it be easier if this were a `Set[string]`?

vMaroon replied:
This affects the comments above on destinations matching logic (implemented).

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+// resolution matches the workload in the placement decision spec.
+func workloadMatchesPlacementDecisionSpec(placementDecisionSpecWorkload *v1alpha1.DownsyncObjectReferences,
+	objectIdentifierToKeyMap map[string]*util.Key, gvkGvrMapper util.GvkGvrMapper) bool {
+	// check lengths
```
If PlacementDecision used a flat structure as in #1222 then this length calculation would be just one use of `len()`.

vMaroon replied:
The structure was inherited from SyncerConfig (@nirrozenbaum).

MikeSpreitzer replied:
Yes, I know.
This is only an FYI, not a blocker.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+	objectIdentifierToKeyMap map[string]*util.Key, gvkGvrMapper util.GvkGvrMapper) bool {
+	for _, clusterScopeDownsyncObjects := range placementDecisionSpecClusterScope {
+		for _, objName := range clusterScopeDownsyncObjects.ObjectNames {
+			gvr := schema.GroupVersionResource(clusterScopeDownsyncObjects.GroupVersionResource)
+			gvk, found := gvkGvrMapper.GetGvk(gvr)
+
+			if !found {
+				return false // assume obj not mapped
+			}
```
This code should be hoisted out of this loop.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,190 @@
+// to simplify its usage and keep implementation details inaccessible.
+type PlacementResolver interface {
+	// GeneratePlacementDecision returns the placement decision for the given
+	// placement key. If a key does not exist in the cache, it is created.
```
What cache is this comment talking about? I do not see that explained.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,190 @@
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+// PlacementResolver abstracts the functionality needed from a placement resolver,
+// to simplify its usage and keep implementation details inaccessible.
```
The comment on this type needs to give a brief outline of what the functionality is; the existing comment is more of a report on the reason this type exists --- which is not bad, but the reader needs to also get a brief description of what this thing is.

The comments on the methods below use mixed terminology. Some speak of "the cache", others speak of "the maintained placement decision".

If I understand correctly, a brief accurate description would be something like the following.

> A PlacementResolver holds a collection of placement decisions. The collection is indexed by `placementKey string`, the resolver does not care what the strings are. Decisions can be updated, externalized, and compared with an external representation.

vMaroon replied:
What do you think about

>// A PlacementResolver holds a collection of placement resolutions.
// The collection is indexed by placementKey string, the resolver does not
// care what the strings are, but a string should be a placement's unique
// identifier. The resolution for a given key can be updated,
// exported and compared to the PlacementDecision representation.

vMaroon replied:
@MikeSpreitzer 
All comments until this point have been addressed. The first batch in the commits referenced in a similar comment above, and this 2nd batch in commits: [addressed mike's 2nd batch of comments in PR 1681, until comment refrencing this commit](https://github.com/kubestellar/kubestellar/pull/1681/commits/ee81ae6b654d52fd635bd30ecd420fa8318c3bce) and [updated resolution comparison functions documentation](https://github.com/kubestellar/kubestellar/pull/1681/commits/55cb1cb811ea64a20c74c7e3b93e206cd89ef1c6).

Outdated comments below have also been addressed.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+	sync.RWMutex
+
+	// map key is GVK/namespace/name as outputted by util.key.GvkNamespacedNameKey()
+	objectIdentifierToKeyMap map[string]*util.Key
```
If the keys in this map used _resource_ instead of _kind_ then there would be less converting between the two. For example, `func (resolver *placementResolver) GeneratePlacementDecision` would not need the ability to return an error.

vMaroon replied:
Assuming you are not referring to refactoring the util.Key struct to include resource name but are suggesting extending the value of this map to be {Key, resource}:
While this would save conversions between the two in GeneratePlacementDecision, the conversions would be paid in UpdateObject (or further up its stack). I'm not sure this is worth it.

If what you are suggesting is refactoring util.Key itself, I believe this can be done in a follow-up work if you don't deem it a necessity for current PR.

MikeSpreitzer replied:
I am not suggesting making Key hold redundant information.

My comment is an FYI, not a blocker. In fact, I would probably prefer the change to be a separate PR. The idea is that the pervasive focus on Kind rather than resource leads to more places in the code needing gvr/gvk mapping than if the pervasive focus were on resource.

vMaroon replied:
I completely agree. 

Just for information: the reasoning for not changing this (and some other choices in the code) is to stay as close as possible to the current implementation and its flow - whether good or bad - since we dont want to propose a full re-implementation of placement. When this PR goes through, this and other changes will be better backed.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+			gvk, found := gvkGvrMapper.GetGvk(gvr)
+
+			if !found {
+				return false // assume obj not mapped
```
The reader wonders why it is valid to simply make an assumption here. If I understand correctly, what is really going on here is that it is always safe to return `false`. The reader needs an explanation, probably centered at the interface definition.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,190 @@
+type PlacementResolver interface {
+	// GeneratePlacementDecision returns the placement decision for the given
+	// placement key. If a key does not exist in the cache, it is created.
+	GeneratePlacementDecision(placementKey string) (*v1alpha1.PlacementDecisionSpec, error)
```
I think the reader should be told something about what errors might arise here. In a controller it is important to know whether an error is indicating a problem that might be cured by a retry later or cannot be cured by a retry later.

If I understand correctly, the only error that can arise here is from a missing entry in a `util.GvkGvrMapper`. Since there is a lot of asynchrony going on here, that is something that actually can happen and is not the result of any user or code doing anything wrong; it _is_ something that is likely to be cured by a retry later.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,190 @@
+	GeneratePlacementDecision(placementKey string) (*v1alpha1.PlacementDecisionSpec, error)
+	// ComparePlacementDecision compares the given placement decision spec
+	// with the maintained placement decision for the given placement key.
+	ComparePlacementDecision(placementKey string,
```
I notice that _this_ method is not defined to be able to return an error. And yet, I see that the implementation _can_ run into trouble with a missing gvk/gvr mapping entry. The reader should be told that this method can return a false negative.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,190 @@
+	// objects data for the given placement key.
+	// If a key does not exist in the cache, it is created.
+	// The returned bool indicates whether the placement resolution was changed.
+	UpdateObjects(placementKey string, obj runtime.Object) (bool, error)
```
The name of this method uses the plural "objects" but the method actually takes only a single object.

The word "update" is very general and vague. I think it would be helpful to be clearer, saying instead that this is an "add".

vMaroon replied:
NoteObject was chosen.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,363 @@
+	resolution.RLock()
+	defer resolution.RUnlock()
+
+	workload := v1alpha1.DownsyncObjectReferences{}
```
I do not see anything setting the `WorkloadGeneration` field.

vMaroon replied:
WIP, I think I should address it properly in a follow up PR. It came AFTER the majority of the code was written, and it requires some refactoring in order for it not to just be set on every update.

vMaroon replied:
@MikeSpreitzer actually, for correctness sake, I will add a simple implementation that tracks generations but makes them local to the controller, and not the CR (which can be argued is correct anyway) - what do you think about that?

MikeSpreitzer replied:
Please open an issue for it and update #1692

vMaroon replied:
Implemented [here](https://github.com/kubestellar/kubestellar/pull/1681/commits/6218a3c7691582cf2ebe17cd2636daa3391ea5e2).

Note that this makes the generation local to the controller, which is unavoidable unless we want to sync state on load (fill the resolutions with the content of the existing placement-decisions, before syncing objects). I'll open an issue.

MikeSpreitzer replied:
I am not clear on what you mean by "while the controller stabilizes" and "sync generation/content on load". Let's pursue those in review of that code.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,190 @@
+	// UpdateDestinations updates the maintained placement's
+	// destinations data for the given placement key.
+	// If a key does not exist in the cache, it is created.
+	UpdateDestinations(placementKey string, destinations []string)
```
"Update" is a very general and vague word. There is a more specific semantic here, right? Wouldn't "SetDestinations" be a clearer name?

MikeSpreitzer commented on pkg/placement/placement.go:
```diff
@@ -142,29 +161,55 @@ func (c *Controller) requeueAll() error {
-	mObj := obj.(metav1.Object)
-	cObj, err := convertToClientObject(obj)
+func (c *Controller) queuePlacementDecision(name string) error {
+	placementDecisions, err := c.listPlacementDecisions()
```
At this point it is already known what `util.KeyForGroupVersionKindNamespaceName` would return, we do not need to get an actual object in hand.

MikeSpreitzer commented on pkg/placement/placement.go:
```diff
@@ -53,33 +52,53 @@ const (
 }
 
-func (c *Controller) requeueForPlacementChanges() error {
+func (c *Controller) requeueForPlacementChanges(placement *v1alpha1.Placement) error {
```
This func is now doing more than "requeue"; I suggest updating the name.

vMaroon replied:
reverted this function to how it was and moved the new logic to handlePlacement

MikeSpreitzer commented on pkg/placement/placement.go:
```diff
@@ -53,33 +52,53 @@ const (
+
+	// requeue all objects to account for changes in placement.
+	// this does not include placement/placement-decision objects.
+	if err := c.requeueAllObjects(); err != nil {
 		return err
 	}
```
This could be simpler.
```suggestion
	return c.requeueAllObjects()
```

MikeSpreitzer commented on pkg/placement/placement.go:
```diff
@@ -53,33 +52,53 @@ const (
+
+	// requeue all objects to account for changes in placement.
+	// this does not include placement/placement-decision objects.
+	if err := c.requeueAllObjects(); err != nil {
 		return err
 	}
 	return nil
```
This could be simpler.

```suggestion
	return c.requeueAllObjects()
```

vMaroon replied:
I originally left this and a few other similar instances for error wrapping later, but now changing all.

MikeSpreitzer commented on pkg/placement/placement.go:
```diff
@@ -179,40 +224,46 @@ func convertToClientObject(obj runtime.Object) (client.Object, error) {
+	labelKey := util.GenerateManagedByPlacementLabelKey(c.wdsName, placement.GetName())
 	for _, manifest := range list.Items {
-		c.logger.Info("Trying to delete manifest", "manifest name", manifest.Name, "namespace", manifest.Namespace, "for placement", mObj.GetName())
+		c.logger.Info("Trying to delete manifest", "manifest name", manifest.Name,
```
This log message is misleading, there is not necessarily an attempt to delete the ManifestWork here.

vMaroon replied:
This, along with object-wrapping/unwrapping code, will later be deleted as part of finalizing this work.

MikeSpreitzer commented on pkg/placement/selectors.go:
```diff
@@ -72,3 +77,97 @@ func GetKeys(m map[string]string) []string {
 }
+
+// when an object is updated, we iterate over all placements and update placementToResolutionMap that
+// are affected by the update. Every affected placement decision is then queued to by synced.
```
s/by synced/be synced/

MikeSpreitzer commented on pkg/placement/selectors.go:
```diff
@@ -72,3 +77,97 @@ func GetKeys(m map[string]string) []string {
+
+	objMR := obj.(mrObject)
+
+	changedResolutions := map[string]interface{}{}
```
I suggest using `Set[string]` here.

vMaroon replied:
A slice of strings now suffices since placement-decisions can be queued by name, with no need to distinguish between existing and none-existing ones.

MikeSpreitzer commented on pkg/placement/selectors.go:
```diff
@@ -72,3 +77,97 @@ func GetKeys(m map[string]string) []string {
+			// mark as handled
+			handledDecisions[name] = nil
+
+			c.logger.Info("enqueued placement decision", "name", name)
```
The log message should say _why_ .

MikeSpreitzer replied:
I see that now there is no log message at all. When I debug controllers, I sometimes find it very helpful if I can set a `-v` level that causes logging of explanations of _why_ something was enqueued.

vMaroon replied:
To know why without adding more indicators, I changed the code to enqueue the placement-decisions right away and not build a slice of them and enqueue after.

MikeSpreitzer replied:
Thanks. The call from `func (c *Controller) handlePlacement` is not logged. Also note: if you log _after_ enqueuing, because of the lack of timing guarantees it is possible that a worker pulls the item and starts working on it (and logging about it) before the log message about enqueuing gets emitted. It would be less confusing the log _before_ enqueuing.

vMaroon replied:
Done.

MikeSpreitzer replied:
The call from `func (c *Controller) handlePlacement` is still not logged.

MikeSpreitzer commented on pkg/placement/selectors.go:
```diff
@@ -72,3 +77,97 @@ func GetKeys(m map[string]string) []string {
+	// at a later stage in the pipeline a worker will realize that the object does not exist
+	// and will create it
+	for name, _ := range changedResolutions {
+		if _, found := handledDecisions[name]; found {
```
FYI, https://github.com/kubernetes/apimachinery/blob/v0.28.2/pkg/util/sets/set.go#L122 is a thing.

vMaroon replied:
A slice of strings now suffices since placement-decisions can be queued by name, with no need to distinguish between existing and none-existing ones.

vMaroon commented on pkg/placement/controller.go:
```diff
@@ -413,13 +440,24 @@ func (c *Controller) reconcile(ctx context.Context, key util.Key) error {
 
 	// avoid further processing for keys of objects being deleted that do not have a deleted object
 	if isBeingDeleted(obj) && key.DeletedObject == nil {
 		return nil
```
isBeingDeleted cannot be true for a key marked for creation. Although this is technically not enforced anywhere, but assumed. I'll add a comment if you agree.

Hopefully what you suggested somewhere below can get rid of IsBeingCreated all-together.

MikeSpreitzer replied:
I assume you are referring to some refactoring that has already been done; I see the `IsBeingCreated` field no longer exists in a `util.Key`.

vMaroon commented:
@MikeSpreitzer thank you for reviewing, all your comments have been addressed.

MikeSpreitzer commented on pkg/crd/crds.go:
```diff
@@ -43,7 +44,10 @@ import (
 
 // CRDs to apply
-var crdNames = map[string]bool{"placements.edge.kubestellar.io": true}
+var crdNames = []string{
+	"placementdecisions.edge.kubestellar.io",
+	"placements.edge.kubestellar.io",
+}
```
Why not do the set construction once here rather than repeatedly inside `filterCRDsByNames`?

```suggestion
var crdNames = sets.New(
	"placementdecisions.edge.kubestellar.io",
	"placements.edge.kubestellar.io",
)
```

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,144 @@
+		obj = util.EmptyUnstructuredObjectFromKey(key)
+	}
+
+	unstructuredObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("failed to convert runtime.Object to unstructured.Unstructured")
+	}
```
This could be confined to an `else` clause of the preceding `if` then , avoiding a type cast in the `then` case (since the only thing that is needed below is the `unstructuredObj`, which `util.EmptyUnstructuredObjectFromKey` produces directly).

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,144 @@
+	// update or create placement decision
+	unstructuredPlacementDecision, err := placementDecisionToUnstructuredObject(pd)
+	if err != nil {
+		return fmt.Errorf("failed to update or create placement decision: %w", err)
```
This error message should be more specific about what went wrong, particularly since the caller is going to wrap with another copy of the same words ("failed to update or create placement decision").

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,144 @@
+
+func setOwnerReference(ownerObj runtime.Object, placementDecision *v1alpha1.PlacementDecision) {
+	// set the placement as the only owner-reference of the placement-decision
+	ownerRef := metav1.NewControllerRef(ownerObj.(metav1.Object), ownerObj.GetObjectKind().GroupVersionKind())
```
Note to self: If the garbage collector is going to delete the PlacementDecision then this controller does not need to. Does it?

vMaroon replied:
Indeed. From what I understood, setting controller to true means setting different owner-references would be blocked, correct?

So when a placement is deleted, the relevant placement-decision is deleted by the garbage collector, and that's it. This controller doesn't have any placement-decision deleting logic, it only deletes the resolution when handling the placement deletion event.

MikeSpreitzer replied:
I don't know about preventing other OwnerReferences (and you did not qualify that statement regarding whether the others have Controller set).

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,144 @@
+	pd.Spec = *placementDecisionSpec
+
+	// set owner reference
+	//	get owner placement object
+	placement, err := c.getPlacementByName(pd.GetName())
+	if err != nil {
+		return fmt.Errorf("failed to get placement %s: %w", pd.GetName(), err)
+	}
```
Since there is already a `PlacementDecisionSpec`, this implies that the corresponding `Placement` has already been read. Why read it again instead of remembering the information needed for the OwnerReference?

vMaroon replied:
I thought you were previously referring to why I'm making an API call to read the placement instead of using the lister's cache, since the placement has been read for sure at a previous step.

Two problems:
* If the spec was just made in-memory (for initial creation), it wouldnt have an owner-reference in the current code, which can be solved by adding one but then we are left with problem 2. 
* Dont we need to verify that the ownerReference is correct and was not modified by anyone?

If you think it's sufficient to set the ownerReference when creating the object (it would still require reading the placement though, but once), I can do that.

MikeSpreitzer replied:
Let's back up and talk about PlacementResolver. The methods NoteObject, RemoveObject, and SetDestinations are only called with a Placement in hand. I am surprised by the comment "If a key does not exist in the cache, it is created" on the GeneratePlacementDecision method. What goes wrong if instead of creating empty content for an absent key, this method returns a `nil` value for the `*v1alpha1.PlacementDecisionSpec`? If a Placement has not been seen, or has been seen _and_ deleted since then, it seems reasonable that this controller should not be trying to write a PlacementDecision value.

MikeSpreitzer replied:
Also: why enqueue a distinct item to make this controller work on a PlacementDecision? Wouldn't it be more natural --- and concurrency-safe --- to do the work on a Placement's decision while syncing a reference to that Placement object?

vMaroon replied:
For RemoveObject I've updated it to not create an empty resolution if none is associated with the placementKey. 

For GeneratePlacementDecision, I admit I didn't really try to simulate all scenarios to figure out what is safe and what isn't, and went for the safe option of: if someone is requesting a generated spec for a resolution that doesn't exist, then the spec should be empty.
Though what I just said can be enforced by the caller of Generate, so I'll implement your suggestion.

MikeSpreitzer replied:
As I understand it, the idea is that syncing a reference to a PlacementDecision is supposed to look in the PlacementResolver to pull out what has been stashed there while syncing other objects and make the external representation in the apiserver(s) match what the PlacementResolver said. When the PlacementResolver has nothing for a given Placement name, this means that the apiserver(s) should hold no corresponding PlacementDecision object.

Think carefully about concurrency and the PlacementResolver. Is it possible that goroutine A calls DeleteResolution for a given placement and then later goroutine B calls NoteObject while working with older information? Yes, it is. That's a bug.

vMaroon replied:
Solved in https://github.com/kubestellar/kubestellar/pull/1681/commits/8af23c51daeb8c1267271dbf5e0badee87bd6a76

The solution is to stop allowing these functions to create resolutions if missing, and to sync up empty resolutions for existing placements on controller load. The alternative was to implement some form of resolution validation that allows for syncing.

vMaroon replied:
Done

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,144 @@
+				Version:  pd.GetObjectKind().GroupVersionKind().Version,
+				Resource: util.PlacementDecisionResource,
+			}).Create(context.Background(), unstructuredPlacementDecision, metav1.CreateOptions{})
+			if err != nil {
```
In the success case it would be better to log a "Created" message and return than to fall through to logging an "Updated" message.

vMaroon replied:
Are you suggesting I only log creation?

vMaroon replied:
Misunderstood previously (and previous comment on this I guess), fixed

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,349 @@
+		gvk, found := gvkGvrMapper.GetGvk(gvr)
+
+		if !found {
+			return false // assume obj not mapped
```
"assume" is still here

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,349 @@
+// updateObject adds/deletes an object to/from the resolution.
+// The return bool indicates whether the placement resolution was changed.
+// This function is thread-safe.
+func (resolution *placementResolution) updateObject(obj runtime.Object) (bool, error) {
```
I would rename this to "noteObject" too.

MikeSpreitzer commented on pkg/util/keys.go:
```diff
@@ -75,3 +120,23 @@ func GenerateObjectInfoString(obj runtime.Object) string {
+
+func KeyIsForPlacementDecision(key Key) bool {
+	if key.GVK.Group == v1alpha1.SchemeGroupVersion.Group &&
+		key.GVK.Version == v1alpha1.SchemeGroupVersion.Version &&
```
I know that in-place upgrades are not in our current level of ambition, but even so: what goes wrong if you omit the version check?

vMaroon replied:
Addressed all comments written before (time-wise) this one.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -63,6 +68,11 @@ func (resolution *placementResolution) noteObject(obj runtime.Object) (bool, err
 	}
 
+	if exists {
+		// the object is already in the resolution but something internal changed
```
Something _internal_ to _what_? I think it would be clearer to s/something internal/the object content/

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -217,6 +217,7 @@ func (resolver *placementResolver) createResolution(placementKey string) *placem
 	placementResolution := &placementResolution{
 		objectIdentifierToKey: make(map[string]*util.Key),
 		destinations:          sets.New[string](),
+		workloadGeneration:    1,
```
What if this process, or the node running it, crashes and later there is a replacement process?

vMaroon replied:
Indeed, for now keeping it like this for correctness under the assumption of friendly environment friendly user. Later on this either has to be synced on load, or this field has to be reevaluated.

MikeSpreitzer replied:
I do not know what you mean by "load". If you want to defer this then please open an Issue for it and update #1692

vMaroon replied:
done #1708

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,149 @@
+	var unstructuredObj *unstructured.Unstructured
+
+	if obj, err := c.getObjectFromKey(key); err != nil {
+		unstructuredObj = util.EmptyUnstructuredObjectFromKey(key)
```
What if the error is not a NotFound error?

vMaroon replied:
the other possibility is that the lister is not found. I think this should be addressed as part of #1690

MikeSpreitzer replied:
This error could be any error returned by the lister. Is NotFound the only error that the lister might return?

vMaroon replied:
Handled explicitly.

MikeSpreitzer commented on pkg/placement/controller.go:
```diff
@@ -387,12 +411,20 @@ func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// special handling for placement-decision resource as it is the only
+	// resource that is queued directly as key, without necessarily first
+	// existing as an object.
+	if util.KeyIsForPlacementDecision(key) {
```
This controller could check for `KeyIsForPlacement` and `KeyIsForCRD` up here too.

vMaroon replied:
They both require obj to be set, which is done inside the block below this line. I think this should be refactored in the work after, what do you think?

MikeSpreitzer replied:
OK with me.

MikeSpreitzer commented on pkg/placement/controller.go:
```diff
@@ -325,6 +335,20 @@ func (c *Controller) enqueueObject(obj interface{}, skipCheckIsDeleted bool) {
 }
 
+func (c *Controller) enqueuePlacementDecision(name string) {
+	c.workqueue.Add(util.Key{
```
FYI, `workqueue.AddAfter` is a thing. This controller could [Nagle](https://en.wikipedia.org/wiki/Nagle%27s_algorithm).

MikeSpreitzer commented on pkg/placement/controller.go:
```diff
@@ -420,6 +449,12 @@ func (c *Controller) reconcile(ctx context.Context, key util.Key) error {
 	}
 
+	if err := c.updateDecisions(obj); err != nil {
+		utilruntime.HandleError(fmt.Errorf("error reconciling object: %s", err))
```
Is it really true that no error possible here is worth a retry? The comment on `updateDecisions` should be explicit about this.

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,149 @@
+		var ok bool
+		unstructuredObj, ok = obj.(*unstructured.Unstructured)
+		if !ok {
+			return fmt.Errorf("failed to convert runtime.Object to unstructured.Unstructured")
```
It was not a _conversation_ that was attempted, only a _recognition_ (technically "type assertion", as written in the comment).

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,149 @@
+	// calculate if the resolved decision is different from the current one
+	if !c.placementResolver.ComparePlacementDecision(placementIdentifier, &placementDecision.Spec) {
+		// update the placement decision object in the cluster by updating spec
+		if err = c.updateOrCreatePlacementDecision(placementDecision, generatedPlacementDecisionSpec); err != nil {
```
FYI. A false negative from `ComparePlacementDecision` due to lack of gvk/gvr mapping will likely cause an error to be returned here for the same reason. I think it would be better to make `ComparePlacementDecision` return an error in that case and save the doomed work here.

vMaroon replied:
I think I'll keep it as it currently is, as this also falls under something that would be resolved by reworking the Key to include object Resource. Is that ok with you?

MikeSpreitzer replied:
yes.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,368 @@
+	objectIdentifierToKey map[string]*util.Key
+	destinations          sets.Set[string]
+
+	// workloadGeneration is a local counter that is changed whenever a mapped
```
"mapped"? Do you mean "matched"? or "listed" or "relevant"?

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,368 @@
+
+// noteObject adds/deletes an object to/from the resolution.
+// The return bool indicates whether the placement resolution was changed.
+// This function is thread-safe.
```
Doesn't this remark belong in the PlacementResolver interface too?

vMaroon replied:
Assuming you only refer to "This function is thread-safe", I've added it to the whole interface (all functions are thread-safe).

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,368 @@
+	_, exists := resolution.objectIdentifierToKey[formattedKey]
+
+	// avoid further processing for keys of objects being deleted that do not have a deleted object
+	if isBeingDeleted(obj) || key.DeletedObject != nil {
```
FYI. `util.KeyForGroupVersionKindNamespaceName` never makes a Key with `DeletedObject != nil`

MikeSpreitzer commented on pkg/placement/placement.go:
```diff
@@ -53,24 +52,35 @@ const (
 	// handle requeing for changes in placement, excluding deletion
 	if !isBeingDeleted(obj) {
+		// update placement resolution destinations since placement was updated
+		list, err := ocm.ListClustersBySelectors(c.ocmClient, placement.Spec.ClusterSelectors)
```
Every caller of `ocm.ListClustersBySelectors` would be better off if that func returned a Set of strings rather than a slice of strings.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,225 @@
+	// SetDestinations updates the maintained placement's
+	// destinations data for the given placement key.
+	// If a key does not exist in the cache, it is created.
+	SetDestinations(placementKey string, destinations []string)
```
FYI: in such situations a comment about mutability is helpful. In this case an accurate statement would be `since the start of a call, nothing mutates the given slice`.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,368 @@
+}
+
+// setDestinations updates the destinations list in the resolution.
+// This function is thread-safe.
```
The interface should also have comments about concurrency and mutability.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,368 @@
+	}
+
+	// the following optimize the building of the workload by maintaining pointers to the object-wrapper structs
+	// that are key'd by GVR
```
FYI: "keyed"

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,368 @@
+	nsObjectsLocationInSlice map[string]int) {
+	if nsdObjects, found := namespaceScopeDownsyncObjectsMap[gvr]; found {
+		// GVR mapping is found, check NS mapping
+		gvrAndNSKey := util.KeyFromGVRandNS(gvr, key.NamespacedName.Namespace) // key for ns mapping inside
```
FYI: This key construction is needed if `!found` too. I would hoist this above the `if`.

MikeSpreitzer replied:
Job half done. The reduant call to `util.KeyFromGVRandNS` is still there.

vMaroon replied:
Missed the one that was not referred to in this comment originally, updated.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,368 @@
+
+// destinationsMatch returns true if the destinations in the resolution
+// match the destinations in the placement decision spec.
+// The function performs one pass over both slices, sequentially.
```
"both slices" is no longer accurate

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,368 @@
+		gvk, found := gvkGvrMapper.GetGvk(gvr)
+
+		if !found {
+			return false // if not found then not mapped and not in resolution
```
Or maybe there was a match when the relevant objects were read but the relevant CRD and all objects whose resource (Kind) was defined by it have been deleted since then and the CRD deletion has been processed by this controller but the object's key is still in the resolution's map (set) of object keys because there is no synchronization that would prevent that.

vMaroon replied:
Good point, this adds more backing to why the key should be reworked (or at least the key I use in the resolution). This issue would be resolved when the said rework is done, therefore I'll leave it as-is for now, ok?

MikeSpreitzer replied:
OK

MikeSpreitzer commented on pkg/util/keys.go:
```diff
@@ -58,9 +77,35 @@ func KeyForGroupVersionKind(group, version, kind string) string {
 }
 
+// Create a string key in the form group/version/resource or version/resource if the group is empty
+func KeyForGroupVersionResource(group, version, resource string) string {
```
FYI: It is confusing that some methods named "KeyFor..." return a Key and others named the same way return a string.

vMaroon replied:
Such naming was already there, I'll open an issue for the rework i keep referring to, which should contain renamings to these functions if they remain.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,225 @@
+
+// A PlacementResolver holds a collection of placement resolutions.
+// The collection is indexed by placementKey string, the resolver does not
+// care what the strings are, but a string should be a placement's unique
```
FYI: Saying that a string "should be" something _here_ suggests that it matters to the PlacementResolver. I think that is false, the comment about strings being a unique identifier for a Placement belongs higher in the stack.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,225 @@
+// care what the strings are, but a string should be a placement's unique
+// identifier. The resolution for a given key can be updated,
+// exported and compared to the PlacementDecision representation.
+type PlacementResolver interface {
```
The comment on this type would be a good place to put a remark about concurrency that covers all the interface's methods. You might also say something about (im)mutability of method parameters, if you think you have something common to say here.

vMaroon replied:
Added, note that this makes the mutability comment on SetDestinations (in the interface) redundant, so removing from there.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,225 @@
+	// GeneratePlacementDecision returns the placement decision for the given
+	// placement key. If a key does not exist in the cache, it is created.
+	// This function can fail if any object's GVR fails to be retrieved given
+	// its GVK, by the mapper shared with the placement-controller.
```
FYI: The definition of this interface does not say anything about a mapper shared with the placement-controller, does not even say anything about a placement-controller at all.

vMaroon replied:
I don't think it's the resolver's responsibility to dictate whether to retry or not. The resolver states what can cause errors, and the reader should decide - depending on his use - whether to retry or not. Is this not right?

MikeSpreitzer replied:
Yes, I see that is done adequately for the resolver.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,225 @@
+	//
+	// - The same is true for every selected object.
+	//
+	// It is possible to output a false negative due to a temporary state of
```
So that "iff" above is not accurate. "only if" would be accurate.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,225 @@
+	// given placement key. If the object is being deleted, it is removed from
+	// the resolution if exists.
+	//
+	// If a key does not exist, it is created.
```
(A) Since the caller is providing the key, it certainly already exists. I suppose this comment is about associating a resolution with the given key.

(B) What is the value in creating an empty resolution and associating it with the key in this case? Why not just return `false`?

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,225 @@
+// exported and compared to the PlacementDecision representation.
+type PlacementResolver interface {
+	// GeneratePlacementDecision returns the placement decision for the given
+	// placement key. If a key does not exist in the cache, it is created.
```
What is "the cache"? This remark should continue to use the terms used in the type comment on this interface.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,225 @@
+// its GVK, by the mapper shared with the placement-controller.
+func (resolver *placementResolver) GeneratePlacementDecision(placementKey string) (
+	*v1alpha1.PlacementDecisionSpec, error) {
+	resolver.RLock() // lock for reading map
+	placementResolution, exists := resolver.placementToResolution[placementKey]
+	resolver.RUnlock() // unlock after accessing map
+
+	if !exists {
+		placementResolution = resolver.createResolution(placementKey)
+	}
```
FYI: I see several repetitions of this sequence of statements, and also a variant that does not create if `!exists`. I usually DRY this out by replacing

```
placementResolution, exists := resolver.placementToResolution[placementKey]
```

with

```
placementResolution, _ := resolver.get(placementKey, true)
```

and replacing

```
	resolver.RLock() // lock for reading map
	placementResolution, exists := resolver.placementToResolution[placementKey]
	resolver.RUnlock() // unlock after accessing map
```

with

```
placementResolution, exists := resolver.get(placementKey, false)
```

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,225 @@
+	placementDecisionSpec, err := placementResolution.toPlacementDecisionSpec(resolver.gvkGvrMapper)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to create placement-decision-spec for placement %v: %w",
```
FYI: I would use the actual type name PlacementDecisionSpec here.

MikeSpreitzer commented on pkg/placement/placement.go:
```diff
@@ -53,24 +52,35 @@ const (
 //     from clusters.
 func (c *Controller) handlePlacement(obj runtime.Object) error {
-	placement := obj.DeepCopyObject()
+	placement, err := runtimeObjectToPlacement(obj.DeepCopyObject())
```
Why two deep copies here?

MikeSpreitzer commented on pkg/placement/placement.go:
```diff
@@ -53,24 +52,35 @@ const (
 //  3. for updates on label selectors, re-evaluate if existing objects should be removed
 //     from clusters.
 func (c *Controller) handlePlacement(obj runtime.Object) error {
```
This PR makes the existing comment on this func inadequate.

MikeSpreitzer commented on pkg/placement/placement.go:
```diff
@@ -124,10 +134,10 @@ func runtimeObjectToPlacement(obj runtime.Object) (*v1alpha1.Placement, error) {
 // this is useful for when a new placement is added
 // or a placement is updated
-func (c *Controller) requeueAll() error {
+func (c *Controller) requeueAllObjects() error {
```
FYI: So maybe "Workload" would be better than "All" in this func's name? What about CRD objects?

vMaroon replied:
This and the above comment would change the existing flow, which I do not intend to change until we're set for disabling existing transport-logic and enabling the new one, which would either be the final phase of this PR or a new PR.

MikeSpreitzer commented on pkg/placement/selectors.go:
```diff
@@ -72,3 +75,50 @@ func GetKeys(m map[string]string) []string {
+			if resolutionUpdated := c.placementResolver.RemoveObject(placement.GetName(), obj); resolutionUpdated {
+				// enqueue placement-decision to be synced since object was removed from its placement's resolution
+				c.enqueuePlacementDecision(placement.GetName())
+				c.logger.V(4).Info("enqueued placement-decision %s for syncing due to the removal of an " +
```
FYI: I would use the type name PlacementDecision and I would identify the workload object.

vMaroon replied:
Oops, fixed.

MikeSpreitzer commented on pkg/placement/selectors.go:
```diff
@@ -72,3 +75,50 @@ func GetKeys(m map[string]string) []string {
+		if resolutionUpdated {
+			// enqueue placement-decision to be synced since an object was added to its placement's resolution
+			c.enqueuePlacementDecision(placement.GetName())
+			c.logger.V(4).Info("enqueued placement-decision %s for syncing due to the addition of an " +
```
Info takes message and key-value pairs, not format and arguments. You forgot the PlacementDecision's name.

MikeSpreitzer replied:
Or you could just include the GVR and the `klog.KSObj` as two key-value pairs.

vMaroon replied:
Fixed the problem (similar to previous comment). Regarding the object identification using `util.GenerateObjectInfoString`, for now im keeping it unified with the rest of the code. This can be a part of the Key refactoring since that function lives there for some reason.

nirrozenbaum commented:
I've created an issue to flatten the CRD structure as suggested by @MikeSpreitzer - #1702. 
I think this is not very critical for now and we should first complete this PR and the other changes in renaming api group and CRDs to avoid conflicts.
we would probably do it in phase1 (after the completion of the above), it's a minor change.

vMaroon commented:
@MikeSpreitzer again thank you for the thorough review. I've addressed all comments, and created issue  #1703 to document the rework/refactoring I referred to multiple times throughout the review.

MikeSpreitzer commented on pkg/placement/placement-resolver.go:
```diff
@@ -0,0 +1,217 @@
+// care what the strings are. The resolution for a given key can be updated,
+// exported and compared to the PlacementDecision representation.
+// All functions in this interface are thread-safe, and no method parameter
+// gets mutated.
```
By anyone, ever? Unlikely to be true. Slices and sets are constructed incrementally, thus mutated. Maybe you want to say something like the following.

> Nothing mutates any method parameter after it is passed in a call to one of these methods.

vMaroon replied:
> Nothing mutates any method-parameter during a call to one of these methods.

MikeSpreitzer commented on pkg/placement/placement.go:
```diff
@@ -43,34 +43,56 @@ import (
 	// handle requeing for changes in placement, excluding deletion
 	if !isBeingDeleted(obj) {
+		// update placement resolution destinations since placement was updated
+		list, err := ocm.ListClustersBySelectors(c.ocmClient, placement.Spec.ClusterSelectors)
```
"list" was never a good name for that variable, something more semantic would have been better. Now it is an even worse name, since the data type is no longer a list.

vMaroon replied:
> FindClustersBySelectors

MikeSpreitzer replied:
The variable is still named "list".

vMaroon commented:
@MikeSpreitzer addressed all new comments.

MikeSpreitzer commented on pkg/placement/placement-decision.go:
```diff
@@ -0,0 +1,160 @@
+	if !c.placementResolver.ResolutionExists(key.NamespacedName.Name) {
+		// if a resolution is not associated to the placement-decision's name
+		// then the placement has been deleted, and the placement-decision has
+		// been garbage collected. We can safely ignore this.
```
It is not guaranteed that the PlacementDecision has already been GCed, only that it eventually will be.

vMaroon replied:
To be sure, you're commenting on the accuracy of the text, and not the correctness of the code, right?

MikeSpreitzer replied:
Yes.

MikeSpreitzer commented on pkg/placement/placement-resolution.go:
```diff
@@ -0,0 +1,366 @@
+	resolution.Lock()
+	defer resolution.Unlock()
+
+	resolution.destinations = destinations.Clone()
```
Making a copy here works.
FYI, this package could be more efficient if it were the callers responsibility to provide a `Set[string]` that never gets mutated henceforth.

MikeSpreitzer commented:
The conversation in https://github.com/kubestellar/kubestellar/pull/1681#pullrequestreview-1857466196 does not need to block rebasing, but I would like to see it finished.

MikeSpreitzer commented:
The comment in https://github.com/kubestellar/kubestellar/pull/1681#issuecomment-1922074820 has not been addressed; that also does not need to block rebase.

MikeSpreitzer commented:
The comment in https://github.com/kubestellar/kubestellar/pull/1681#discussion_r1477173528 has not been addressed; can be done after rebase.

vMaroon commented:
@MikeSpreitzer addressed comments - except for the first two of the last 3 PR comments.

Rebased onto main.

vMaroon commented:
On the failing test, when running locally:

> ➜  kubestellar git:(merge) ✗ wait-for-cmd 'expect-cmd-output '\''kubectl --context imbs1 get manifestworks -n cluster1; kubectl --context imbs1 get manifestworks -n cluster2'\'' '\''wc -l | grep -wq 8'\'''
NAME                                       AGE
addon-addon-status-deploy-0                98m
v1-namespace--nginx                        12m
appsv1-deployment-nginx-nginx-deployment   12m
NAME                                       AGE
addon-addon-status-deploy-0                98m
v1-namespace--nginx                        12m
appsv1-deployment-nginx-nginx-deployment   12m
^C%                                                                                                                                              
➜  kubestellar git:(merge) ✗ wait-for-cmd 'expect-cmd-output '\''kubectl --context imbs1 get manifestworks -n cluster1; kubectl --context imbs1 get manifestworks -n cluster2'\'' '\''wc -l'\'''             
NAME                                       AGE
addon-addon-status-deploy-0                99m
v1-namespace--nginx                        13m
appsv1-deployment-nginx-nginx-deployment   13m
NAME                                       AGE
addon-addon-status-deploy-0                99m
v1-namespace--nginx                        13m
appsv1-deployment-nginx-nginx-deployment   13m
^C%                                        

Which you can see hangs, but running manually:

> ➜  kubestellar git:(merge) ✗ (kubectl --context imbs1 get manifestworks -n cluster1; kubectl --context imbs1 get manifestworks -n cluster2) | wc -l 
       8
➜  kubestellar git:(merge) ✗ (kubectl --context imbs1 get manifestworks -n cluster1; kubectl --context imbs1 get manifestworks -n cluster2) | wc -l | grep -wq 8
➜  kubestellar git:(merge) ✗

MikeSpreitzer commented on pkg/placement/controller.go:
```diff
@@ -237,6 +247,11 @@ func (c *Controller) run(workers int) error {
 
+	// populate the PlacementResolver with entries for existing placements
+	if err := c.populatePlacementResolverWithExistingPlacements(); err != nil {
+		return fmt.Errorf("failed to populate the PlacementResolver for the existing placements")
```
`err` should make it into the returned error.

vMaroon replied:
Fixed.

MikeSpreitzer commented:
@vMaroon : thanks. I reviewed it all again, found just one new thing, noted just above.

MikeSpreitzer commented:
I have asked GitHub to retry the failed workflow job.

MikeSpreitzer commented:
The test/e2e/multi-cluster-deployment succeeds for me too when I run it on my laptop.

MikeSpreitzer commented:
To investigate failures in CI, the workflows can be extended to print a bunch of relevant state.  See https://github.com/kubestellar/kubestellar/blob/main/.github/workflows/docs-ecutable-user-quickstart-test.yml#L71-L95 for an example (specific to the legacy architecture, of course).

MikeSpreitzer commented on test/e2e/common/setup-shell.sh:
```diff
@@ -25,7 +25,7 @@ function wait-for-cmd() (
         if (($wait_counter > 10)); then
             echo "Failed to ${cmd}."
-            kubectl -n wds1-system logs -l control-plane=controller-manager
+            echo $(kubectl -n wds1-system logs -l control-plane=controller-manager)
```
Why wrap the `kubectl` command into an `echo` command?

vMaroon replied:
I was expecting a log, so thought it's not working for some reason. Which wasn't the issue.

MikeSpreitzer commented:
@vMaroon : in your remark "A branch that changes retries to 35 failed 5/8 times. (Again retries are not really exhausted)", what do you mean by "retries are not really exhausted"?

MikeSpreitzer commented:
If we need 60 retries to get a reliable test, then so be it.

vMaroon commented:
@MikeSpreitzer what I mean is that when it works, it works right away - not that the increase from 36 to 50 made it work.

MikeSpreitzer commented:
Both tests failed with 200 retries. Just now I asked GitHub to re-run the failed workflow jobs. We will see what that produces.

We need to debug this. We need the workflow to unconditionally dump the controller logs and object contents. We need the placement controller to log enough. We need OCM to log enough.

pdettori commented:
I got a bit lost in this thread, I understand that there is instability in the test results and it may take a lot of retries to get the test to complete. Are these tests reliable when running locally? If that is that case, could it be that there are not sufficient/reliable resources when running in the CI/CD VM ? @francostellari has evaluated that to run the 3 kind clusters + KF / KS + wds1 / imbs1 takes about extra 3 GB of memory.

vMaroon commented:
@pdettori when running locally, everything worked 100% of the time. In CICD, sometimes it worked, sometimes did not. 
But there is a bug in the controller's setup (::run) that in main's code leads to a lot of failures on workers wake, and in my code to death.

The bug is here:
    

     // ensure CRDs are installed before starting up
	 if err := crd.ApplyCRDs(c.dynamicClient, c.kubernetesClient, c.extClient, c.logger); err != nil {
              return err
	 }

	// Get all the api resources in the cluster
	apiResources, err := c.kubernetesClient.Discovery().ServerPreferredResources()
	if err != nil {
		// ignore the error caused by a stale API service
		if !strings.Contains(err.Error(), util.UnableToRetrieveCompleteAPIListError) {
			return err
		}
	}


Between these two, nothing makes sure that the CRDs have been digested by the apiserver.
Therefore, if the Placement CRD does not make it into the apiResources list, the workers are woken into failures (since workload object processing requires listing placements, and there is no lister for placements yet in this case), until the Placement CRD is reconciled by a worker.

In the new code, there is a new function that is a part of ::run (`populatePlacementResolverWithExistingPlacements`) that expects the existence of the placements lister in the listers map. Therefore in the scenario above, it kills the launch.

pdettori commented:
@vMaroon thank you for the explanation, now I understand. Is there anything we could better in the  `waitForCRDAccepted` to ensure that the API is in discovery ? perhaps use the discovery client at that point and poll until it shows up there?

pdettori commented:
@vMaroon - FYI - I found and fixed also a race condition in the startup of the status controller that may also crash the controller, but you are not yet hitting that it seems  - see https://github.com/kubestellar/kubestellar/pull/1707/files#diff-4752b8f7ccbed6c0bbc49182b308efcbd91e01dbe87f2d612b24616b4e0f20d4

MikeSpreitzer commented:
I have drafted a fix for the problem @vMaroon identified, see https://github.com/vMaroon/kubestellar/pull/8 .  I am running the workflows on it now.

vMaroon commented:
@MikeSpreitzer I have tested that, it doesn't work. I've also tested waiting until listing CRDs covers the API CRDs [as such](https://github.com/vMaroon/kubestellar/commit/3611e0528eee135299bbbdffd546a3f1dcbeb200), also doesn't work.

My next planned attempt is to wait until `c.kubernetesClient.Discovery().ServerPreferredResources()` covers them.

MikeSpreitzer commented:
The workflow runs on https://github.com/MikeSpreitzer/kcp-edge-mc/tree/more-1681 succeeded.  I just asked GitHub to retry them.

The mere existence of the CRD objects is not enough, that is implied by the fact that their NamesAccepted Condition is not enough.

vMaroon commented:
Apologies @MikeSpreitzer my last comment was wrong, I did test it earlier and it worked https://github.com/vMaroon/kubestellar/actions/runs/7805371893/job/21289436982, if I recall correctly, I dropped it due to what I now understand is a misread of it being inferior to NameAccepted.

It is time to sleep.

pdettori commented:
@vMaroon @MikeSpreitzer great find! I should have waited for `Established` indeed, can't remember now why I have used `NameAccepted`

MikeSpreitzer commented:
See also https://github.com/vMaroon/kubestellar/pull/9 . We do not have to merge that change into @vMaroon 's `merge` branch, I could open a PR against `main` for it --- after _this_ PR merges. Or before, if @vMaroon wants to trim the workflow changes out of this PR.

MikeSpreitzer commented:
@pdettori : did you clone the placement controller outline --- including the submission of the CRD object definitions to the apiserver --- from somewhere? I am wondering if something about this problem should go back upstream.

pdettori commented:
@MikeSpreitzer the general outline was largely based on the [sample controller](https://github.com/kubernetes/sample-controller). The crd check was not taken from a specific example, I just used a pattern I was familiar with based on `wait.PollUntilContextCancel` to periodically poll for an object and check the status. The issue is that I looked at the wrong condition.

ezrasilvera commented:
/lgtm
/approve
